### PR TITLE
Make --rock-affix-params work for rock.gemm

### PR DESCRIPTION
**Pull request by krzysz00**
1. Make most of the tuning code take a RockGemmWrapperInterface instance and not a ConvContext
2. Extend RockGemmWrapperInterface to allow for that
3. Rename GemmContext to GemmSize and make it carry the g dimension as well because I can't believe I didn't do that
1. Create the .cpp file for the RockGemmWrapperInterface I'd forgotten about
5. Do all the reorginazation needed to make this work

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: I think we can drop dir -- it is not being used.

File: mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:207]: Here also, we can drop the dir as it is not being used.

File: mlir/lib/Dialect/Rock/Tuning/ConvContext.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Tuning/ConvContext.cpp:50]: I think we dont need this opType as part of the ConvContext after this change.
My quick investigation suggests that only uses of populeConvContext remains in ConvToGemm (after this PR) where the caller knows the direction and even more I dont see ctx.opType is being used.

**Review by manupak — COMMENTED** [review]
Changes looks good overall.

My only comment here stems from the fact we dont need to pass around a None coming from obtain obtainConvDirection(...).

To elaborate further, I think after the change populateConvContext gets limited to ConvToGemm (rightfully so...). In there, all the callers know the direction and I dont think ctx.opType is ever used. Therefore, I think we can safely remove opType from ConvContext which removes the need/scope for passing around a None in the case of rock.gemm.

File: mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: I left it in there because we might eventually have cases where backward data etc. have different sets of good heuristic parameters - we had those before, IIRC.

**Comment by krzysz00** [comment]
@jerryyin Re having the convolution direction passed in to the "get heuristic parameter list" function ... do you think we should drop that parameter or leave it in just in case?

File: mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp
Comment (jerryyin) [inline mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp:371]: How are you able to move template definition into a cpp without running into linker issues? I'm a bit surprised that ... Unless this is used only inside of this cpp but not anywhere else. Is that the case?

File: mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp
Comment (manupak) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: How about we pass the Op directly instead as we are expanding this pass to rock.gemm now ?

Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: I'd love to do that but `Conv2DGenerator` calls this for the backward weight padding kernel stuff

Comment (manupak) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: Hmmm.. I thought getTuningParameters is only called under affix params pass.

Just to clarify, my suggestion was to replace the dir with the Op (that already captures each convop corresponding to the direction as well as gemm op) here, assuming the clients of this function within affix params pass.

IIRC, there is a different function to obtain kernel counts, workspace, etc that also scrolls these things but it was not getTuningParameters

Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: The call is in `calculatePaddingKernelSize()` in `Conv2dGenerator::needExtraPadBwdWeight`

If it weren't for that one call, I would've made this take the op just like the rest of the system

Comment (manupak) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:552]: ah yes... I remember now.

File: mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp
Comment (jerryyin) [reply mlir/lib/Dialect/Rock/Tuning/GridwiseGemmParams.cpp:207]: Looks like we can indeed drop the direction. But I'm catching up here... we used to initialize that heuristic parameters based different direction as well. I seem to notice that `orderInitPrams()` does its job on deciding based on amount of padding. Do we do any special ordering based on conv direction as well?

Thinking out loud: It might be that we used a consistent set of heuristic init parameters against all 3 conv directions, which is how you are able to simplify the logic here?

An orthogonal topic, it looks like xdlops and nonxdlops has some unevenness on int8, and I suspect we'd need to bring over a different set of tuning parameters on int8 nonxdlops as well. It may be a rocm x.x ticket.

**Review by jerryyin — APPROVED** [review]

File: mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp:371]: I moved it to the only file that used it

**Comment by krzysz00** [comment]
@jerryyin We don't - and haven't for a while - done any selection based on forward/bwd data/bwd weight. We use the same set of heuristic parameters for every direction, and only switch sets based on data type.

We don't necessarily need to do that: we could have different parameter sets for different ops. The question for this PR is "do we keep that unused parameter (op direction) in the API to get the heuristic list?"

**Comment by jerryyin** [comment]
> do we keep that unused parameter (op direction) in the API to get the heuristic list?

I don't think it matters. It is very easy to add the direction back, so we can remove the parameter and add it back later. 

Part of the reason, as I can see that we want to keep it is when we need different set of tuning parameter for different types, like int8.

I don't see this blocking this PR so approved it. I'll leave it to you to decide.

**Comment by krzysz00** [comment]
Given that this is a rather important PR and that we're liable to refactor all this anyway, I'm fine keeping it as is

**Pull request merged**
